### PR TITLE
wasmparser: Apply clippy suggestions + add clippy CI job

### PR DESCRIPTION
**Pull request by Robbepop**
This PR adds a `clippy` CI job for `wasmparser` that can be extended to other crates of this workspace in later PRs.
Also the PR applies all `clippy` suggestions to the `wasmparser` crate.

### CODE REVIEW

File: crates/wasmparser/src/validator.rs [hunk 1]
```diff
@@ -1792,7 +1791,7 @@ mod arc {
                 return None;
             }
             debug_assert!(Arc::get_mut(&mut self.arc).is_some());
-            Some(unsafe { &mut *(&*self.arc as *const T as *mut T) })
+            Some(self.assert_mut())
```
Comment (Robbepop) [inline crates/wasmparser/src/validator.rs:1794]: The downside of this change is that it introduces a runtime check.
The ideal replacement was [`Arc::get_mut_unchecked`](https://doc.rust-lang.org/alloc/sync/struct.Arc.html#method.get_mut_unchecked) but that's still unstable as of now.

**Comment by Robbepop** [comment]
I don't understand why the tests are failing. The `wasmparser` tests work fine locally.

File: crates/wasmparser/src/validator.rs (hunk 1 above)
Comment (alexcrichton) [reply crates/wasmparser/src/validator.rs:1794]: This needs to stay as it was, otherwise this is a mutally recursive function that never returns.

**Comment by Robbepop** [comment]
@alexcrichton Thanks for the review! Applied your suggestions.

**Comment by Robbepop** [comment]
@alexcrichton Would you be interested for a similar PR for the `wast` crate? I'd be down to fix `clippy` lints there as well as adding it to the CI when everything is cleaned up.

File: crates/wasmparser/src/validator.rs (hunk 1 above)
Comment (Robbepop) [reply crates/wasmparser/src/validator.rs]: Applied your suggestion and tests are running now again, too. :tada:

File: crates/wasmparser/src/validator.rs [hunk 2]
```diff
@@ -1792,7 +1791,19 @@ mod arc {
...
-            Some(unsafe { &mut *(&*self.arc as *const T as *mut T) })
+            Some(
+                // # Safety
+                //
+                // This operation is safe since the API asserts that
+                // `self.owned` is only ever `true` if there is just
+                // one owner of the `Arc`.
+                // As soon as `Arc::get_mut_unchecked` we should
+                // make use of it instead.
+                #[allow(clippy::cast_ref_to_mut)]
```
Comment (alexcrichton) [inline crates/wasmparser/src/validator.rs:1802]: I would prefer to have no `#[clippy]` attributes in this crate, can this be removed?

File: crates/wasmparser/src/validator.rs [hunk 3]
```diff
@@ -1792,7 +1791,19 @@ mod arc {
...
             debug_assert!(Arc::get_mut(&mut self.arc).is_some());
-            Some(unsafe { &mut *(&*self.arc as *const T as *mut T) })
+            Some(
+                // # Safety
+                //
+                // This operation is safe since the API asserts that
+                // `self.owned` is only ever `true` if there is just
+                // one owner of the `Arc`.
+                // As soon as `Arc::get_mut_unchecked` we should
+                // make use of it instead.
```
Comment (alexcrichton) [inline crates/wasmparser/src/validator.rs:1801]: This `# Safety` comment doesn't really add anything over the structure of the code that's already here, so I don't think it should be added.

File: crates/wasmparser/src/validator.rs (hunk 2 above)
Comment (Robbepop) [reply crates/wasmparser/src/validator.rs:1802]: How are we going to satisfy the `clippy` CI then?
Are those `#[clippy]` attributes okay in the crate root? Less ideal though ...

Comment (alexcrichton) [reply crates/wasmparser/src/validator.rs:1802]: We don't run clippy in CI in this repository, so there's nothing to satisfy.

Comment (Robbepop) [reply crates/wasmparser/src/validator.rs:1802]: Is there a reason for that? With this PR the `wasmparser` crate could profit from having a `clippy` CI job at hand. In fact this PR adds a `clippy` CI job for `wasmparser`.

Comment (alexcrichton) [reply crates/wasmparser/src/validator.rs:1802]: Personally I do not believe that clippy adds value to a project. That's my personal opinion, others may feel differently. I don't really want to debate the merits of clippy here at this time, my point is that we don't run clippy and that is not something I would like to change off-hand in a PR.

**Comment by Robbepop** [comment]
I applied your review comments:

- Removed `clippy` CI job for `wasmparser`.
- Removed `Safety` comment.
- Remove attribute to silence a `clippy` warning for unsafe code.

**Pull request merged**